first version of linear regression to compute co-efficients

sudansohan opened the pull request:
Hi Aksel,
Could you please review if I am correctly accessing the buffers across different functions? I will update the reduction to use hierarchical implementation. I still need to write the verification function, etc. So, you do not need to merge this one yet.
Bests,
Sohan

illuhad commented:
Yes, this would work although you don't necessarily have to pass the buffers as references. `buffer` and many other SYCL objects work similarly to a reference counted pointer, so you can just pass them by value into the functions. A copy of a buffer object will still refer to the same underlying memory region.

sudansohan commented:
Thanks for the review. This also solved the problem that I discussed during our Skype call.

sudansohan commented:
added validated version of linear regression to find co-efficient that fits the line. Renamed the earlier version (of linear regression that computes error) to remove the ambiguity. 

Aksel, I would like to merge this one if you are ok with it.

illuhad commented on single-kernel/lin_reg_coeff.cpp:
```diff
@@ -0,0 +1,194 @@
+int main(int argc, char** argv)
+{
+  BenchmarkApp app(argc, argv);
+  app.run<LinearRegressionCoeffBench<float>>();
```
This should be surrounded with `if(app.shouldRunNDRangeKernels())` since this benchmark uses hierarchical parallel for. This will allow us to disable the benchmark when the `--no-ndrange-kernels` command line option is used.

sudansohan replied:
hi Aksel, thanks for the quick comment. The current version does not use the hierarchical parallel for, but I will do that.

sudansohan commented:
hi Aksel, 
please have a look. I would like to merge if you are okay.

illuhad commented on single-kernel/lin_reg_coeff.cpp:
```diff
@@ -0,0 +1,196 @@
+      auto in1 = input1_buf.template get_access<s::access::mode::read>(cgh);
+      auto in2 = input2_buf.template get_access<s::access::mode::read>(cgh);
+ 
+       // Use discard_write here, otherwise the content of the host buffer must first be copied to device
```
By the way, when `output_buf` hasn't been written before on the host, a good SYCL runtime will recognize that a data transfer is unnecessary even for a regular `write` access mode since both host and device buffer are in the same state :)

illuhad commented on single-kernel/lin_reg_coeff.cpp:
```diff
@@ -0,0 +1,196 @@
+                  local_mem[local_mem_index] = local_mem[local_mem_index] + local_mem[local_mem_index + stride];
+              }
+
+              item.barrier(s::access::fence_space::local_space);
```
I think the reduction from the reduction benchmark may be more efficient because of memory bank conflicts and because it saves one barrier. We don't need to have the most efficient reduction of all time, but maybe we can agree on one way to formulate reductions such that the benchmarks reuse the same patterns?

sudansohan replied:
I was also thinking of this. Maybe we can move some commonly used functions to common file and just call them when needed. No need to replicate code.

illuhad replied:
Yes, I think that's a good idea.

sudansohan replied:
Good then, we can discuss this in our next call and do it!

illuhad commented on single-kernel/lin_reg_coeff.cpp:
```diff
@@ -0,0 +1,196 @@
+    T ss_xx = reduce(output_buf);
+
+    T mean_x = reduce(input1_buf)/args.problem_size;
+    T mean_y = reduce(input2_buf)/args.problem_size;
```
Are we able to merge some of the kernels here for efficiency?

sudansohan replied:
Probably possible, but I am not sure the effect on caching, etc. because a thread needs to make access to two different arrays. Maybe AoS can help. But it requires more effort :)

sudansohan commented:
are you going to merge with the master? lin_reg.cpp was renamed to lin_reg_error.cpp

illuhad commented:
can you resolve the merge conflicts?
